visualizer/draw/bubbleprof-ui.js

```diff
@@ -353,12 +329,81 @@ class BubbleprofUI extends EventEmitter {
       }
       currentUI = currentUI.selectNode(layoutNode, animationQueue)
       if (layoutNode.node.uid === aggregateNode.uid) {
-        executeAnimationQueue(animationQueue)
+        animationQueue.execute()
         return currentUI
       }
     }
   }
 
+  setupHistory () {
+    // Mark the current history entry (on page load) as the initial one.
+    // There is a `window.history.length` property, but it includes the entries
+    // _before_ the current page (eg. the github issue that linked to this
+    // visualization) so we can't use it here.
+    window.history.replaceState({ initial: true }, null, window.location.hash || '')
+
+    this.on('navigation', ({ to, silent }) => {
+      // Only update history if this navigation was not caused by history.
+      if (!silent) {
+        this.pushHistory(to.getHash())
+      }
+
+      this.backBtn.isHidden = !this.hasHistoryEntries()
+      this.backBtn.draw()
+    })
+
+    window.addEventListener('popstate', (event) => {
+      const { hash } = event.state
+      if (this.topMostUI) {
+        // Close stack frames when moving away from their node.
+        if (this.topMostUI.selectedDataNode) {
+          this.topMostUI.clearFrames()
+        }
+      }
+
+      // Prepare the jump we need to make.
+      const jump = () => {
+        this.historyAnimationQueue = new AnimationQueue('history')
+        this.historyAnimationQueue.onComplete = () => {
+          this.historyAnimationQueue = null
+        }
+
+        if (hash) {
+          this.topMostUI.jumpToHash(hash, this.historyAnimationQueue)
+        } else {
+          // If we don't have a hash, we're navigating to the original UI.
+          this.topMostUI.traverseUp(null, {
+            silent: true,
+            animationQueue: this.historyAnimationQueue
+          })
+        }
+      }
+
+      // If we have don't have a history animation queue, create one and jump
+      // immediately.
+      if (!this.historyAnimationQueue) {
+        jump()
```

Comment:
That is when a non-history animation is going on, correct? Eg. clicking into a cluster node and then using the history before the animation completes.
I punted on that because it seemed a bit more tricky to handle non-history animations, since they can be started from any sublayout as well as the original UI. But probably the `historyAnimationQueue` could be replaced with a `this.originalUI.currentAnimationQueue` property and all UIs could use that in the same way to queue animations globally, so it might not be as difficult as I thought.

Reply:
Yes, that's right. Sounds like a good approach, treating animations the same as much as possible regardless of origin